src/java/org/apache/cassandra/cql3/QueryProcessor.java

```diff
@@ -747,7 +761,17 @@ public static ResultMessage.Prepared storePreparedStatement(String queryString,
                                                             DatabaseDescriptor.getPreparedStatementsCacheSizeMB(),
                                                             queryString.substring(0, 200)));
         MD5Digest statementId = computeId(queryString, keyspace);
-        Prepared previous = preparedStatements.get(statementId, (ignored_) -> prepared);
+        // Caffeine key collision of concurrent invalidation and computation has to be avoided DSP-24330
+        Prepared previous;
+        try
+        {
+            while (preparedStatementsUpdateLock.putIfAbsent(statementId, false) != null);
```

Comment:
This is the mirror image of the guard in `evictPrepared`, the thread executing the access to the `preparedStatements` won't let the thread with the same statement id execute the eviction until the access to `preparedStatements` is done.

The small side effect of this synchronization that I see is that currently if we have two queued threads that execute the `storePreparedStatement` with the query that has the same hash (likely the same query) then the latter will have to wait until the former finishes reading from cache. I don't know how often that happens and whether it would be a major issue though.

Reply:
Considering they rely on a MD5 key I would consider this possible but rare. Even when it happened it would only have to wait an extra ` put` and `remove  operations. That I don't like but as I said this is the least workable solution I could come up with.

Reply:
> The small side effect of this synchronization that I see is that currently if we have two queued threads that execute the storePreparedStatement with the query that has the same hash (likely the same query) then the latter will have to wait until the former finishes reading from cache. I don't know how often that happens and whether it would be a major issue though.


Yeah, this does happen.  Think of client applications, they often prepare all their statements at startup.  It's not unusual that a fleet (cluster) of applications all start at once.  This means not just the queue wait time, but multiple threads all thrashing cpu spinning…

Given that concurrent `storePreparedStatement` don't create a problem, what about…
```
while (!Boolean.TRUE.equals(preparedStatementsUpdateLock.putIfAbsent(statementId, Boolean.FALSE)));
```

and then above for `evictPrepared` (line 241)
```
            while (!Boolean.FALSE.equals(preparedStatementsUpdateLock.putIfAbsent(id, Boolean.TRUE)));
```

Reply:
Actually that makes sense. 2 concurrent stores will get through and then it's caffeine's decision on how to handle that. Yep I like a lot.

Reply:
Good example @michaelsembwever and a sensible suggestion as well! It gives us a semantic where many concurrent writes OR concurrent evictions (synchronized internally by the thread-safe cache) are possible but concurrent conflicting operations are mutually exclusive.

One more aspect of this synchronization that comes to my mind is fairness, using the concurrent map the way it's used now to synchronize the CS allows a situation where one type of operation starves another (in theory).

Reply:
> One more aspect of this synchronization that comes to my mind is fairness, using the concurrent map the way it's used now to synchronize the CS allows a situation where one type of operation starves another (in theory).

in `storePreparedStatement` 
```
while (!Boolean.TRUE.equals(preparedStatementsUpdateLock.putIfAbsent(statementId, Boolean.FALSE)))
{
   LockSupport.parkNanos(1);
}
```
might help fairness (and/or prioritise `invalidate`), with minimal overhead.

More info in 
- https://arxiv.org/abs/1305.5800
- https://issues.apache.org/jira/browse/CASSANDRA-15922 

But, I doubt this is needed, i would want to see micro-benchmarking to see it did what was intended.

Reply:
> I doubt this is needed

Yes, I don't think we need to worry that much about the fairness of these operations, I don't expect this area will suffer from high contention that lasts long.

> i would want to see micro-benchmarking to see it did what was intended.

I remember @bereng mentioned the other day he did some micro-benchmarking already.

Reply:
No jmh. Just plain stress load locally on my laptop and nothing showed up.